### PR TITLE
[WIP] enable build of cross target components

### PR DESCRIPTION
**Pull request by rahku**

### CODE REVIEW

File: CompileDefinitions.cmake [hunk 1]
```diff
@@ -0,0 +1,177 @@
+if (WIN32)
```
Comment (janvorli) [inline CompileDefinitions.cmake]: Please use only lowercase letters for filenames. That is a convention that we have picked for coreclr at the beginning.

**Comment by rahku** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by janvorli** [comment]
@rahku I completely disagree with the refactoring. All of the added .cmake files for sources doesn't seem to help anything and I see it as an unnecessary complication. The only place where it makes sense to me is for the functions and definitions in the root CMakeLists.txt since you need it for the crossplat folder.
I like the previous model where each folder has a single CMakeLists.txt.

File: CMakeLists.txt [hunk 2]
```diff
@@ -689,211 +125,49 @@ if(WIN32)
...
+  add_compile_options(/Zl) # omit default library name in .OBJ
+endif(WIN32)
 
+#-------------------------------------
 # Include directory directives
-
+#-------------------------------------
 # Include the basic prebuilt headers - required for getting fileversion resource details.
-include_directories("src/pal/prebuilt/inc")
+include_directories("${CLR_DIR}/src/pal/prebuilt/inc")
```
Comment (janvorli) [inline CMakeLists.txt]: The CLR_DIR is not needed here, since it is the current directory.

**Comment by rahku** [comment]
currently the way it works is you start from the root cmakelists.txt and build everything according to the folder hierarchy. In order to build cross target components I do not want to build everything. For example to build crossgen I only intend to build src/jit/crossgen from jit folder and nothing else in jit folder. In order to do that I create a common jit.cmake which is used by all subfolders. In this way I can separately build jit/crossgen. I happy to hear any other suggestion you might have for this.

File: crosscomponents/cmakecmd.bat [hunk 3]
```diff
@@ -0,0 +1 @@
+cmake -DCMAKE_USER_MAKE_RULES_OVERRIDE=F:\coreclrGit\coreclr\crosscomponents\windows-compiler-override.txt -DCLR_CMAKE_HOST_ARCH=x64 -DCLR_CMAKE_TARGET_ARCH=arm64 -G "Visual Studio 14 2015 Win64"  %1
```
Comment (janvorli) [inline crosscomponents/cmakecmd.bat]: It seems this file should not be part of the change (guessing from the hardcoded path)

Comment (rahku) [inline crosscomponents/cmakecmd.bat]: yes I still need to do some clean up. Only CMakeLists.txt in this folder will remain ...rest everything will go away.

**Comment by janvorli** [comment]
It seems you can easily handle that by ifs in the CMakeLists.txt - you would define a variable for the crossbuild and if that one was set, you would not add the subfolders you don't want to add for crossbuild. That would preserve the very important aspect of the cmake - the hierarchical way, Before your change, when I want to know what include folders and defines are set for a specific folder, I can just check the hierarchy. With your way, you include folders in an arbitrary manner and it is very hard to track the hierarchy of the build.

**Comment by rahku** [comment]
I did think about that and didn't like it. Reasons below:
1. lots of if(NOT DEFINED CROSS_COMPILE). Because we build less for cross target so most will be actually under the if. Didn't like the clutter.
2. For any change you need to know about cross_compile. Because by default you need to put it under if(not defined cross_compile). Didn't like that it wasn't simple for normal (non-cross target) build.
The current method is not that arbitrary. It will still be hierchical for normal(non-cross) build. I need to fix the places that don't do that right now. Only for cross-target build it will not have hierarchy but a flat list of dirs to build.

**Comment by janvorli** [comment]
Another way would be to create a variable with a list of targets that should be built for the cross target and modify the add_library_clr and add_executable_clr functions to check if we are building for the cross target if the library / exe name is in the list (using list(FIND ...) command) and if it is not, just skip the body of the functions.
That way both builds would use the same unmodified CMakeLists.txt hierarchy and all the stuff would be centralized.
At the time cmake generates the makefiles / VS projects, it would just find the targets that it needs to build and generate the makefiles / VS projects only for those.
Would this work for you?

**Comment by rahku** [comment]
There are certain libs & executables which build without being wrapped under these wrappers. Those would get built for both builds. Also I believe what I have is much more simpler to understand and maintain. Like I said for non-cross builds the hierarchy will still be present. I will fix that. Only thing that will change is that folders will have a *.cmake files with source listing besides CMakeLists.txt.

**Comment by rahku** [comment]
Pattern for creating .cmake is already used in repo and I don't see anything wrong with that

**Comment by janvorli** [comment]
I disagree that it is much simpler to understand and maintain. Instead of a small local change in one file, you needed to modify 90 files. And tearing the dac components build out of the hierarchy while it should follow the same hierarchy as the matching wks components is also unfortunate. The way you have chosen forces you to set sources dir at each specific build type of a component separately (e.g. src/md/runtime/crossgen/CMakeLists.txt, src/md/runtime/dac/CMakeLists.txt, src/md/runtime/dbi/CMakeLists.txt) instead of nicely using the hierarchy.
I consider the hierarchical directory tree following nature of the cmake build files with the hierarchical dependencies its most important concept that was making it so much nicer than the msbuild projects that are a nightmare to follow. And it goes away with this change. We still build it hierararchically, but we had to add a lot of glue instead of the automatic dependencies.

**Comment by rahku** [comment]
You will not need to modify 90 files. It will still be one local change. That is the idea behind this change. dac will not be teared outside hierarchy. I will fix that. This is the way it will be...taking md/runtime as an example.
src/md/runtime/CmakeLists.txt -> will add all the sub_dirs including dac 
src/md/runtime/runtime.cmake -> all source listing & common cmake config
src/md/runtime/dac/cmakelists.txt -> will include ../runtime.cmake & add_library for runtime_dac
src/md/runtime/wks/cmakelists.txt -> will include ../runtime.cmake & add_library for runtime_wks
src/md/runtime/crossgen/cmakelists.txt -> will include ../runtime.cmake & add_library for runtime_cross

This way dac & crossgen can be built as part of hierarchy as they are added in src/md/runtime/CmakeLists.txt. But they can also be build directly without following the hierarchy by just adding src/md/runtime/crossgen directory in crosscomponents/CmakeLists.txt 

So if you have to make any change that should go in runtime.cmake and just that file and it will reach every subfolder.

**Comment by janvorli** [comment]
But you still either cannot depend on the hierarchical inheritance of include folders, defines and options or you have to duplicate them carefully in your crosscomponents/CMakeLists.txt so that they match the values accumulated in the hierarchy. Which I can see you are doing. And that seems error prone since any addition of such thing in the hierarchy will need to be reflected in the crosscomponents/CMakeLists.txt and it is easy to miss by a developer.
I have experimented with the way I have suggested (local change in the add_library_clr, add_executable_clr and install_clr) and it seems it works as expected.
Please see the diff here:
https://github.com/dotnet/coreclr/compare/master...janvorli:crosstarget-build-experiment?expand=1
It builds and installs the targets your change was building.
One more additional change would be to exclude building of corerun and coreconsole that are not created using add_executable_clr and install_clr, but using the plain add_executable and install. We would probably want to create additional clones of those for targets that don't depend on the etw stuff, but that should be a trivial change at few places.

**Comment by janvorli** [comment]
Btw, as you can see, I didn't have to include all the dependencies of the selected targets, just the targets themselves. The trick is that all targets but the selected ones are excluded from "all targets" list of targets that are always built, but it doesn't prevent cmake from building the right dependencies.

**Comment by janvorli** [comment]
And this is the output folder after the build:
```
 Directory of D:\git\coreclr\bin\Product\Windows_NT.x64.Debug
04/13/2016  04:03 PM    <DIR>          .
04/13/2016  04:03 PM    <DIR>          ..
04/13/2016  04:02 PM    <DIR>          .nuget
04/13/2016  04:03 PM    <DIR>          PDB
04/13/2016  04:01 PM    <DIR>          gcinfo
04/13/2016  04:01 PM    <DIR>          inc
04/13/2016  04:01 PM    <DIR>          lib
04/13/2016  04:03 PM         5,187,072 mscordbi.dll
04/13/2016  04:01 PM         2,461,696 mscorlib.dll
04/13/2016  04:02 PM        16,884,736 mscorlib.ni.dll
04/13/2016  04:03 PM         9,722,368 crossgen.exe
04/13/2016  04:03 PM         2,300,928 sos.dll
04/13/2016  04:03 PM         5,122,048 mscordaccore.dll
               6 File(s)     41,678,848 bytes
               7 Dir(s)  1,145,061,347,328 bytes free
```

**Comment by gruberjuliannama** [comment]
runtime/CmakeLists.txt -> will add all the sub_dirs including dac 
src/md/runtime/runtime.cmake -> all source listing & common cmake config
src/md/runtime/dac/cmakelists.txt -> will include ../runtime.cmake & add_library for runtime_dac
src/md/runtime/wks/cmakelists.txt -> will include ../runtime.cmake & add_library for runtime_wks
src/md/runtime/crossgen/cmakelists.txt -> will include ../runtime.cmake & add_library for runtime_cross

This way dac & crossgen can be built as part of hierarchy as they are added in src/md/runtime/CmakeLists.txt. But they can also be build directly without following the hierarchy by just adding src/md/runtime/crossgen directory in crosscomponents/CmakeLists.txt,,?

**Comment by rahku** [comment]
Thanks Jan for the early feedback. Yes your solution is better. I will open another PR.

**Comment by janvorli** [comment]
Thank you!
